src/python/pants/backend/java/lint/google_java_format/rules.py

```diff
@@ -0,0 +1,174 @@
+import dataclasses
+from dataclasses import dataclass
+from typing import Iterator
+
+from pants.backend.java.lint.google_java_format.skip_field import SkipGoogleJavaFormatField
+from pants.backend.java.lint.google_java_format.subsystem import GoogleJavaFormatSubsystem
+from pants.backend.java.lint.java_fmt import JavaFmtRequest
+from pants.backend.java.target_types import JavaSourceField
+from pants.core.goals.fmt import FmtResult
+from pants.core.goals.lint import LintRequest, LintResult, LintResults
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import Digest, MergeDigests
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.process import BashBinary, FallibleProcessResult, Process, ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import FieldSet, Target
+from pants.engine.unions import UnionRule
+from pants.jvm.jdk_rules import JdkSetup
+from pants.jvm.resolve.coursier_fetch import MaterializedClasspath, MaterializedClasspathRequest
+from pants.jvm.resolve.jvm_tool import JvmToolLockfileRequest, JvmToolLockfileSentinel
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+
+@dataclass(frozen=True)
+class GoogleJavaFormatFieldSet(FieldSet):
+    required_fields = (JavaSourceField,)
+
+    source: JavaSourceField
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipGoogleJavaFormatField).value
+
+
+class GoogleJavaFormatRequest(JavaFmtRequest, LintRequest):
+    field_set_type = GoogleJavaFormatFieldSet
+
+
+class GoogleJavaFormatToolLockfileSentinel(JvmToolLockfileSentinel):
+    options_scope = GoogleJavaFormatSubsystem.options_scope
+
+
+@dataclass(frozen=True)
+class SetupRequest:
+    request: GoogleJavaFormatRequest
+    check_only: bool
+
+
+@dataclass(frozen=True)
+class Setup:
+    process: Process
+    original_digest: Digest
+
+
+@rule(level=LogLevel.DEBUG)
+async def setup_google_java_format(
+    setup_request: SetupRequest,
+    tool: GoogleJavaFormatSubsystem,
+    jdk_setup: JdkSetup,
+    bash: BashBinary,
+) -> Setup:
+    source_files, tool_classpath = await MultiGet(
+        Get(
+            SourceFiles,
+            SourceFilesRequest(field_set.source for field_set in setup_request.request.field_sets),
+        ),
+        Get(
+            MaterializedClasspath,
+            MaterializedClasspathRequest(
+                prefix="__toolcp",
+                lockfiles=(tool.resolved_lockfile(),),
+            ),
+        ),
+    )
+
+    source_files_snapshot = (
+        source_files.snapshot
+        if setup_request.request.prior_formatter_result is None
+        else setup_request.request.prior_formatter_result
+    )
+
+    input_digest = await Get(
+        Digest,
+        MergeDigests([source_files_snapshot.digest, tool_classpath.digest, jdk_setup.digest]),
+    )
+
+    def add_java16_or_higher_options() -> Iterator[str]:
```

Comment:
doing this with inline functions reads quite weirdly (notwithstanding that this is JVM support code :)). You could just as easily do the following, there'd be less indentation, and less surprise for the reader.

```
maybe_java16_or_higher_options: List[str] = []
if jdk:
    maybe_java16_or_higher_options = [
        ...
    ]
```